### PR TITLE
[BBT#541] A new way to install the internal headers

### PR DESCRIPTION
**Pull request by bosilca**
https://bitbucket.org/icldistcomp/parsec/pull-requests/541

Move the install directive in the corresponding CMake file.

Clean the install and the usage of PaRSEC as an internal or external project.

The other 2 commits were minor cleanups to remove typos and compiler warnings.

### CODE REVIEW

File: parsec/CMakeLists.txt [hunk 1]
```diff
@@ -236,9 +244,116 @@ if( BUILD_PARSEC )
...
+    get_target_property(_public_headers parsec PUBLIC_HEADER_H)
+    foreach(_FILE ${_public_headers})
+        get_filename_component(_DIR ${_FILE} DIRECTORY)
+        INSTALL(FILES ${_FILE} DESTINATION ${PARSEC_INSTALL_INCLUDEDIR}/parsec/${_DIR})
+    endforeach()
+    get_target_property(_private_headers parsec PRIVATE_HEADER_H)
+    foreach(_FILE ${_private_headers})
+        get_filename_component(_DIR ${_FILE} DIRECTORY)
+        INSTALL(FILES ${_FILE} DESTINATION ${PARSEC_INSTALL_INCLUDEDIR}/parsec/${_DIR})
+    endforeach()
```
Comment (devreal) [inline parsec/CMakeLists.txt:262]: The indentation is off here, that threw me off while trying to understand what it's doing

**Review by devreal — APPROVED** [review]

File: parsec/CMakeLists.txt [hunk 2]
```diff
@@ -120,6 +108,31 @@ endif( PARSEC_PROF_GRAPHER )
 # Setup targets
 #
 if( BUILD_PARSEC )
+    # As suggested @ https://crascit.com/2016/01/31/enhanced-source-file-handling-with-target_sources/
```
Comment (abouteiller) [inline parsec/CMakeLists.txt:111]: indentation

File: parsec/interfaces/dtd/CMakeLists.txt [hunk 3]
```diff
@@ -9,9 +9,13 @@ if( BUILD_PARSEC )
     DESTINATION ${PARSEC_INSTALL_INCLUDEDIR}/parsec/interfaces/dtd/)
 
   if( PARSEC_WITH_DEVEL_HEADERS )
-    install(FILES
-            ${CMAKE_CURRENT_SOURCE_DIR}/interfaces/dtd/insert_function_internal.h
-            DESTINATION ${PARSEC_INSTALL_INCLUDEDIR}/parsec/interfaces/dtd/)
+    set_property(TARGET parsec
+                 APPEND PROPERTY
+                        PUBLIC_HEADER_H interfaces/dtd/insert_function_internal.h)
+#George
```
Comment (abouteiller) [inline parsec/interfaces/dtd/CMakeLists.txt:15]: stray comment

File: parsec/CMakeLists.txt [hunk 4]
```diff
@@ -234,11 +242,118 @@ if( BUILD_PARSEC )
...
+  get_target_property(_private_headers parsec PRIVATE_HEADER_H)
+  foreach(_FILE ${_private_headers})
+    get_filename_component(_DIR ${_FILE} DIRECTORY)
+    INSTALL(FILES ${_FILE} DESTINATION ${PARSEC_INSTALL_INCLUDEDIR}/parsec/${_DIR})
+  endforeach()
+
+  install(FILES utils/help-mca-param.txt
+          DESTINATION ${PARSEC_INSTALL_DATADIR}/parsec)
+  install(FILES
+          ${CMAKE_CURRENT_SOURCE_DIR}/runtime.h
```
Comment (abouteiller) [inline parsec/CMakeLists.txt:267]: Why aren't these also dealt with `set_property(TARGET parsec PUBLIC_HEADER)`?

File: parsec/CMakeLists.txt [hunk 5]
```diff
@@ -234,11 +242,118 @@ if( BUILD_PARSEC )
...
 
   install(TARGETS parsec
-    EXPORT parsec-targets
-    ARCHIVE DESTINATION ${PARSEC_INSTALL_LIBDIR}
-    LIBRARY DESTINATION ${PARSEC_INSTALL_LIBDIR}
-    PUBLIC_HEADER DESTINATION ${PARSEC_INSTALL_INCLUDEDIR})
+          EXPORT parsec-targets
+          ARCHIVE DESTINATION ${PARSEC_INSTALL_LIBDIR}
+          LIBRARY DESTINATION ${PARSEC_INSTALL_LIBDIR})
+  # Cmake < 3.20 does not support hierarchical PUBLIC and PRIVATE _HEADER,
```
Comment (abouteiller) [inline parsec/CMakeLists.txt:248]: Maybe it could make sense to require cmake >= 3.20, we are already at 3.18 so its a small step.

**Review by abouteiller — CHANGES_REQUESTED** [review]
not tested, just read through

File: parsec/CMakeLists.txt (hunk 4 above)
Comment (bosilca) [reply parsec/CMakeLists.txt:267]: With current cmake I could have used PUBLIC_HEADER with all headers going directly in include. It looks cleaner if the PUBLIC_HEADER_H is only used for headers outside the parsec/CMakeLists.txt, and everything else is installed via INSTALL.

**Review by abouteiller — APPROVED** [review]

**Pull request merged**